src/metrics/handler.py

```diff
@@ -0,0 +1,152 @@
+from datetime import timedelta
+from os.path import join
+from django.conf import settings
+import inspect
+import uuid
+from metrics import utils
+from metrics.utils import ensure
+import requests, requests_cache
+import logging
+
+requests_cache.install_cache(**{
+    'cache_name': join(settings.OUTPUT_PATH, 'db'),
+    'backend': 'sqlite',
+    'fast_save': True,
+    'extension': '.sqlite3',
+    # https://requests-cache.readthedocs.io/en/latest/user_guide.html#expiration
+    'expire_after': timedelta(hours=24 * settings.CACHE_EXPIRY)
+})
+
+def clear_cache():
+    requests_cache.clear()
+
+LOG = logging.getLogger('debugger') # ! logs to a different file at a finer level
+
+def opid(nom=''):
+    "return a unique id to track a set of operations"
+    # ll: pmc--20170101-235959--48029843290842903824930
+    return '--'.join(filter(None, [nom, utils.ymdhms(), str(uuid.uuid4())]))
+
+def fqfn(fn):
+    mod = inspect.getmodule(fn)
+    return '.'.join([mod.__name__, fn.__name__])
+
+def writefile(id, content, fname):
+    path = join(settings.DUMP_PATH, id)
+    ensure(utils.mkdirs(path), "failed to create path %s" % path)
+    path = join(path, fname) # ll: /tmp/elife-metrics/pmc-asdfasdfasdf-482309849230/log
+    open(path, 'w').write(content)
+    return path
+
+#
+#
+#
+
+class NoneObj(object):
+    pass
+
+def ignore_handler(id, err):
+    "does absolutely nothing"
+    pass
+
+def logit_handler(id, err):
+    "writes a comprehensive log entry to a file"
+    # err ll: {'request': <PreparedRequest [GET]>, 'response': <Response [404]>}
+    payload = {
+        'request': err.request.__dict__,
+        'response': err.response.__dict__
+    }
+    fname = writefile(id, utils.lossy_json_dumps(payload), 'log')
+    body = err.response.content
+    fname2 = writefile(id, body, 'body')
+    ctx = {
+        'id': id,
+        'logged': [fname, fname2],
+    }
+    LOG.warn("non-2xx response %s" % err, extra=ctx)
+
+def raise_handler(id, err):
+    "logs the error and then raises it again to be handled by the calling function"
+    logit_handler(id, err)
+    raise err
+
+RAISE, IGNORE, LOGIT = 'raise', 'ignore', 'log'
+HANDLERS = {
+    RAISE: raise_handler,
+    IGNORE: ignore_handler,
+    LOGIT: logit_handler
+}
+DEFAULT_HANDLER = raise_handler
+
+def requests_get(*args, **kwargs):
+    id = kwargs.pop('opid', opid())
+    ctx = {
+        'id': id
+    }
+    handler_opts = kwargs.pop('opts', {})
+
+    try:
+        # TODO: opportunity here for recovery from certain errors
+        resp = requests.get(*args, **kwargs)
+        resp.raise_for_status()
+        return resp
+
+    except requests.HTTPError as err:
```

Comment:
it could be easier to just check `resp.status_code` than to throw an exception with `raise_for_status` (although other exceptions like connection errors will still happen)